src/Microsoft.Framework.FileSystemGlobbing/Internal/PatternContexts/PatternContextRagged.cs

```diff
@@ -81,19 +70,49 @@ public override void PushFrame(DirectoryInfoBase directory)
                 }
             }
 
-            PushFrame(frame);
+            PushDataFrame(frame);
+        }
+
+        public struct FrameData
+        {
+            public bool IsNotApplicable;
+
+            public int SegmentGroupIndex;
+
+            public IList<IPathSegment> SegmentGroup;
+
+            public int BacktrackAvailable;
+
+            public int SegmentIndex;
+        }
+
+        protected IRaggedPattern Pattern { get; }
+
+        protected bool IsStartsWith()
```

Comment:
Moving them back to method simply for the sake of readability since they're all simply method to test a condition. A method looks more straightforward. There is not performance impact per say. 

In terms of the name, how about this:

`IsStartingGroup`
`IsEndingGroup`

And simply remove the `IsContains` since it is used in only one place. There it can test `!IsStartingGroup() && !IsEndingGroup()`